### PR TITLE
Replace QuickESPNow with local library

### LINKED ISSUE #5614: With ESPNow remote and no wifi WLED reboots every 15-20min (fixed in 17.0.0-dev)
### What happened?

Without wifi to connect to the device is rebooting every 15-20min, if the AP it connects to is available then it appears to be stable over multiple days.

Runs fine on 14.  As per https://github.com/wled/WLED/issues/3685#issuecomment-4273014821 this bug is being opened.

possibly related to: https://github.com/wled/WLED/issues/3685

### To Reproduce Bug

Connect the device to a wifi access point then turn off the AP.

### Expected Behavior

Should be stable without wifi.

### Install Method

Binary from WLED.me

### What version of WLED?

WLED_16.0.0-beta_ESP32

### Which microcontroller/board are you seeing the problem on?

ESP32

### Relevant log/trace output

```shell

```

### Anything else?

This is in a VERY isolated area, the wifi it connects to is literally the only wifi around.

It is paired with an ESPNow remote, but during the 15-20 before the reboot no buttons have been pressed.

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by DedeHai**
QuickESPNow is not well suited as it is focused on compatibility rather than speed and resources. 

This PR replaces it with local functions directly based on esp API calls.
From my tests it is faster and more light weight:
- ESP32 shows a saving of ~10k of heap
- ESP8266 shows a saving of ~1.5k of heap

I tested this with a send and receive function both with large packets and bursts of small packets. While large packets work quite well, small (5 bytes) packets tend to get lost on ESP8266 when sent in bursts longer than 6 packets. 
The changes need testing in real applications but bench-top tests are very promising.

edit:
it also saves 2k of flash

## Summary

* **Refactor**
  * Switched ESP‑NOW handling to a built-in broadcast-only driver for initialization, callbacks, sending, and shutdown.
  * Unified AP/STA setup using the same flow.

* **New Features**
  * Added a lightweight `wled-espnow` module providing broadcast-only transmission with in-flight throttling.

* **Bug Fixes**
  * Improved segmented ESP‑NOW packet sending so failures stop further segments and errors are reported consistently.

* **Chores**
  * Cleaned up ESP‑NOW build configuration by removing the prior optional dependency and adjusting feature flags/ignore lists.

### CODE REVIEW

File: wled00/src/dependencies/espnow_wled/espnow_wled.cpp [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+#else  // ESP8266
+  if (esp_now_init() != 0) {
+    DEBUG_PRINTLN(F("ESP-NOW esp_now_init() failed"));
+    return false;
+  }
+  esp_now_set_self_role(ESP_NOW_ROLE_COMBO); // TODO: found no official documentation on this... quickespnow ESP_NOW_ROLE_SLAVE in STA mode and ESP_NOW_ROLE_CONTROLLER in AP mode
+  esp_now_register_recv_cb(_espnowRecvCB);
+  esp_now_register_send_cb(_espnowSentCB);
+  esp_now_add_peer(const_cast<uint8_t*>(BCAST), ESP_NOW_ROLE_COMBO, channel, nullptr, 0);
+  _running = true;
```
[the remark this answers is not in the document]
Comment (DedeHai) [reply wled00/src/dependencies/espnow_wled/espnow_wled.cpp]: the original implementation in quickesp now did not do any checks and has been running for years, it seems this is safe and callback registring does not fail.

File: lib/wled_espnow/wled_espnow.h [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+// ESPnow WLED: Lightweight ESP-NOW driver. Only the API surface actually used by WLED  is implemented.
+// by @dedehai (2026) licensed under EUPL 1.2 license (same as WLED)
+
+#ifndef WLED_DISABLE_ESPNOW
+#ifdef ESP8266
+  #include <espnow.h>
+#else
+  #include <esp_now.h>
+#endif
```
[the remark this answers is not in the document]
Comment (softhack007) [reply lib/wled_espnow/wled_espnow.h:12]: we only have esp32 and esp8266. let's not over-complicate things.

**Comment by DedeHai** [comment]
I have some pending changes, the actual bug the rabbit did not even mention...

File: lib/wled_espnow/wled_espnow.cpp [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+  }
+  if (err == 0) _inFlight++; // ESP_OK == 0 on both platforms
+  else if (_inFlight > 0 && !isretransmit) {
+    uint8_t lastInFlight = _inFlight;
+    delay(2); // wait for a queued message to be sent, found that 2ms is usually enough, dont want to be too cautios (burst send is currently an edge case)
+    // note: delay and general approach might need some tweaking for real world use, based on burst tests sending 16 messages
+    if (_inFlight < lastInFlight) {
+      isretransmit = true; // try once more
+      err = esp_now_send(const_cast<uint8_t*>(BCAST), const_cast<uint8_t*>(data), len);  // A message was sent and the sent callback was called, so we can retry now.
+    }
```
[the remark this answers is not in the document]
Comment (DedeHai) [reply lib/wled_espnow/wled_espnow.cpp]: it should actually be like this:

`      if (err == 0) _inFlight++; // 0 is ESP_OK but is not defined on ESP8266
      else Serial.printf("ESP-NOW send failed with error %d, inflight=%d\n", err, (int)espNow._inFlight);
      return err; `
 
otherwise isretransmit is set to false again at the end of the function.

**Comment by DedeHai** [comment]
@willmmiles question about folder and file structure: I put this in src/dependencies but it feels kind of out of place since it is not an external dependency. I could put it in the main code folder but it is cluttered as it is. Any suggestion on a better folder structure to put such "local library" files?

**Comment by willmmiles** [comment]
> @willmmiles question about folder and file structure: I put this in src/dependencies but it feels kind of out of place since it is not an external dependency. I could put it in the main code folder but it is cluttered as it is. Any suggestion on a better folder structure to put such "local library" files?

I absolutely agree that `src/dependencies` should be reorganized.   PlatformIO suggests that local or vendored libraries should be subfolders in `lib/`.   That's probably a good place if you want to make sure it's treated as quasi-independent library, written with a presumption that it could (in theory) be used outside of WLED (even if that never actually happens).

If it was really up to me, though ... I'd pull ESP-NOW support out of the core entirely and move it to `usermods/esp_now`.  ;)   The library back-end could naturally live there.  It'd be a good test case for finding the places where the module APIs need extension.  (Blockers that I know about offhand: settings UI integration, reading cfg root, notify integration -- I think most other call sites already have hooks available.)

**Comment by willmmiles** [comment]
If you do go with `lib/` , I suggest inverting the name (`wled-EspNow`) and using a namespace of `wled::EspNow` for the functions and objects.  A couple of things have been pointing in favor of a common `wled` root namespace (doesn't conflict with current `class WLED`), and I've got `wled::Log` and `wled::Net` kicking about in prototype code.

**Comment by DedeHai** [comment]
I think making core libraries a usermod puts too much strain on the concept, at least for now.

**Comment by willmmiles** [comment]
> I think making core libraries a usermod puts too much strain on the concept, at least for now.

Oh for sure - there are still some key APIs missing.  I do think that moving things in to modules instead of `#ifdef`s and expanding our set of default modules is the best path for the long run, though.   I'm really quite excited to see integrations like Hue V2, Matter, and FPP all available through the module API instead of adding to our tangled core.

**Comment by netmindz** [comment]
Sounds like a very good idea

**Comment by DedeHai** [comment]
@softhack007  please stop pushing directly to main and go through PRs. it causes things to get reverted upon "merge from main" which can cause issues if I made a change and when updating from main it is suddenly gone again.

**Comment by DedeHai** [comment]
now compiles for V5 on all platforms, tested the latest version on ESP32, C6 and C3, V5 looks good.
Still needs testing on C5, H2, C61. 
If there are no objections I will merge this.

for reference: if anyone want to run my test function, use this code on two devices.

in wled.cpp, replace
`wled::espNow.onDataRcvd(espNowReceiveCB); `
with 
`wled::espNow.onDataRcvd(espNowTestRecvCB);`

and add this just after `handleRemote();`

`if(millis() > 8000)  handleEspNowTest();`

and add the test function above the `void WLED::loop()`


```C++
#ifndef WLED_DISABLE_ESPNOW
// ─────────────────────────
// ESP-NOW send/receive test
// ─────────────────────────

struct EspNowTestMsg {
  uint8_t mac[6];
  uint8_t data[250];
  uint8_t len;
  int8_t  rssi;
  bool    broadcast;
};
static constexpr uint8_t ESPNOW_TEST_RING_SIZE = 16;
static EspNowTestMsg     espNowTestRing[ESPNOW_TEST_RING_SIZE];
static volatile uint8_t  espNowTestHead = 0; // written by recv callback
static uint8_t           espNowTestTail = 0; // written by loop()

// Recv callback — WiFi driver task context, no Serial allowed.
static void espNowTestRecvCB(uint8_t *mac, uint8_t *data, uint8_t len,
                             signed int rssi, bool broadcast) {
  uint8_t next = (espNowTestHead + 1) % ESPNOW_TEST_RING_SIZE;
  if (next == espNowTestTail) return; // ring buffer full — drop
  EspNowTestMsg &slot = espNowTestRing[espNowTestHead];
  memcpy(slot.mac, mac, 6);
  uint8_t copy = (len < 250) ? len : 250;
  memcpy(slot.data, data, copy);
  slot.len       = copy;
  slot.rssi      = (int8_t)rssi;
  slot.broadcast = broadcast;
  espNowTestHead = next; // publish: advance head only after writing slot contents
}

static void handleEspNowTest() {
  static bool          cbRegistered = false;
  static unsigned long lastSend     = 0;
  static unsigned long lastBurst    = 0;
  static uint32_t      sendCounter  = 0;

  if (statusESPNow != ESP_NOW_STATE_ON) return;

  // Register test recv callback once (overrides the production callback).
  if (!cbRegistered) {
    wled::espNow.onDataRcvd(espNowTestRecvCB);
    cbRegistered = true;
    Serial.println(F("ESP-NOW test: recv callback registered"));
  }

  // Drain all queued received messages.
  while (espNowTestTail != espNowTestHead) {
    const EspNowTestMsg &msg = espNowTestRing[espNowTestTail];
    Serial.printf("ESP-NOW RX: mac=%02x:%02x:%02x:%02x:%02x:%02x  rssi=%d  bcast=%d  len=%u  data=",
      msg.mac[0], msg.mac[1], msg.mac[2],
      msg.mac[3], msg.mac[4], msg.mac[5],
      msg.rssi, (int)msg.broadcast, msg.len);
    for (int i = 0; i < msg.len && i < 16; i++)
      Serial.printf("%02x ", msg.data[i]);
    if (msg.len > 16) Serial.print(F("..."));
    Serial.println();
    espNowTestTail = (espNowTestTail + 1) % ESPNOW_TEST_RING_SIZE;
  }

  unsigned long now = millis();

  // Every 5 s: burst-send 18 messages to verify the in-flight cap (max 16).
  // The first 16 should succeed (ret=0), sends 17 and 18 should be rejected (ret=1).
  if (now - lastBurst >= 5000) {
    lastBurst = now;
    lastSend  = now; // suppress the normal heartbeat on this same iteration
    Serial.printf("ESP-NOW burst test (inflight before=%d):\n", (int)wled::espNow._inFlight);
    for (int i = 0; i < 36; i++) {
      uint8_t msg[8] = { 'B', 'S', 'T', (uint8_t)i };
      uint8_t ret = wled::espNow.send(ESPNOW_BROADCAST_ADDRESS, msg, sizeof(msg));
      if (ret != 0) Serial.printf("  send[%02d] rejected (inflight=%d), err=%u\n", i, (int)wled::espNow._inFlight, ret);
      //Serial.printf("  send[%02d] ret=%u  inflight=%d\n", i, ret, (int)wled::espNow._inFlight);
    }
    Serial.printf("inflight=%d\n", (int)wled::espNow._inFlight);
  }
  // Every 1 s: send a heartbeat and show the current in-flight counter.
  else if (now - lastSend >= 1000) {
    lastSend = now;
    sendCounter++;
    // Payload: "HB" magic, 4-byte counter, current inflight count, padding
    uint8_t msg[8];
    msg[0] = 'H'; msg[1] = 'B';
    msg[2] = (sendCounter >> 24) & 0xFF;
    msg[3] = (sendCounter >> 16) & 0xFF;
    msg[4] = (sendCounter >>  8) & 0xFF;
    msg[5] =  sendCounter        & 0xFF;
    msg[6] = (uint8_t)wled::espNow._inFlight;
    msg[7] = 0;
    uint8_t ret = wled::espNow.send(ESPNOW_BROADCAST_ADDRESS, msg, sizeof(msg));
    Serial.printf("ESP-NOW TX: #%lu  ret=%u  inflight=%d\n",
                  sendCounter, ret, (int)wled::espNow._inFlight);
  }
}
#endif // !WLED_DISABLE_ESPNOW
```
